### PR TITLE
Refactor Python gRPC pipeline 

### PR DESCRIPTION
**Pull request by geigerj**
- Introduce a task that renames Python gRPC package declarations
- Use grpcio-tools directly instead of Packman to run codegen
- Move Python gRPC from Packman to MetadataGen

Note that the renaming task does not play nicely with the baseline tests, which do not permit the filesystem to be mocked, so I had to mock a portion of that task in the baseline test. Full coverage is provided through separate unit tests.

Updates #97 

cc: @jmuk @garrettjonesgoogle

### CODE REVIEW

File: test/test_python_package_change_task.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        'import "google/common/common_proto.proto";\n',
+        'Some other text referencing to google.service.v1\n']
+
+    def test__extract_base_dirs(self):
+        mock_proto = mock.mock_open()
+        mock_proto.return_value.__iter__ = lambda _: iter(self._PROTO_FILE)
+        with mock.patch('__builtin__.open', mock_proto, create=True):
+            base_dirs = self._TASK._extract_base_dirs(
+                os.path.join('a', 'test', 'path', 'to', 'google', 'service',
+                             'v1', 'a.proto'))
```
Comment (jmuk) [inline test/test_python_package_change_task.py]: Please add a test case that the path of the workspace itself contains similar pattern, like `/home/foo/workspaces/google/service/v1/googleapis/google/service/v1/a.proto` ?

File: test/test_python_package_change_task.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                         'google/cloud/grpc/other')
+        self.assertEqual(self._TASK._transform('google$service', '$', []),
+                         'google$cloud$grpc$service')
+
+        # Don't transform common protos
+        self.assertEqual(
+            self._TASK._transform('google/common', '/', ['google.common']),
+            'google/common')
+        self.assertEqual(
+            self._TASK._transform('google/uncommon', '/', ['google.service']),
```
Comment (jmuk) [inline test/test_python_package_change_task.py]: I'd like to keep the same common protos list, so using `['google.common']`?

Comment (jmuk) [inline test/test_python_package_change_task.py]: or `google/common` against `['google.service']` might make sense.

File: test/test_pipeline_baseline.py [hunk 3]
```diff
@@ -85,14 +86,16 @@ def _test_error(pipeline_name, language, config, pipeline_kwargs,
         execute_pipeline.main(args)
 
 
-@mock.patch('pipeline.utils.task_utils.run_gradle_task')
+@mock.patch('pipeline.utils.task_utils.get_gradle_task_output')
+@mock.patch('pipeline.tasks.protoc_tasks.PythonChangePackageTask._copy_proto')
```
Comment (jmuk) [inline test/test_pipeline_baseline.py]: Can we just mock the entire `PythonChangePackageTask` (or its `execute`)? That would simplify the setup of the whole test -- you don't need several yaml files in the testcase (and I assume that this task is well tested in another test file).

File: pipeline/tasks/protoc_tasks.py [hunk 4]
```diff
@@ -402,6 +451,48 @@ def execute(self, language, api_name, output_dir, src_proto_path,
...
+    def execute(self, api_name, toolkit_path, descriptor_set, service_yaml,
+                intermediate_package_dir, output_dir,
+                package_dependencies_yaml, package_defaults_yaml, language):
+        service_args = ['--service_yaml=' + os.path.abspath(yaml)
+                        for yaml in service_yaml]
+        pkg_dir = os.path.join(output_dir, 'python', 'grpc-' + api_name)
+
+        # TODO(geigerj): This section temporarily replicates packman behavior.
+        # Instead, these configuration values should be derived from the artman
+        # conifg.
```
Comment (garrettjonesgoogle) [inline pipeline/tasks/protoc_tasks.py]: spelling(conifg)

File: pipeline/tasks/protoc_tasks.py [hunk 5]
```diff
@@ -455,3 +546,122 @@ def execute(self, gapic_api_yaml):
...
+        package_suffix='\\.proto',
+        suffix='";'))
+
+    # TODO (geigerj): add regex for documentation link updates?
+
+    def execute(self, src_proto_path, import_proto_path, common_protos_yaml):
+        with open(common_protos_yaml) as common_protos_file:
+            common_protos_data = yaml.load(common_protos_file)
+
+        # Treate google.protobuf as a common proto package, even though it is
```
Comment (garrettjonesgoogle) [inline pipeline/tasks/protoc_tasks.py]: spelling (Treate)

File: pipeline/tasks/protoc_tasks.py [hunk 6]
```diff
@@ -455,3 +546,122 @@ def execute(self, gapic_api_yaml):
                 continue
             if 'package_name' in go_settings:
                 return go_settings.get('package_name')
+
+
+class PythonChangePackageTask(task_base.TaskBase):
```
Comment (garrettjonesgoogle) [inline pipeline/tasks/protoc_tasks.py]: There is a lot of stuff here for Python - would it make sense to move it to a separate file?

Comment (geigerj) [inline pipeline/tasks/protoc_tasks.py]: Sure, will do in the following commit on this PR

File: pipeline/tasks/protoc_tasks.py (hunk 5 above)
Comment (geigerj) [inline pipeline/tasks/protoc_tasks.py]: Done

File: pipeline/tasks/protoc_tasks.py (hunk 4 above)
Comment (geigerj) [inline pipeline/tasks/protoc_tasks.py]: Done

File: pipeline/tasks/protoc_tasks.py (hunk 6 above)
Comment (geigerj) [inline pipeline/tasks/protoc_tasks.py]: Done

File: test/test_python_package_change_task.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        mock_proto = mock.mock_open()
+        mock_proto.return_value.__iter__ = lambda _: iter(self._PROTO_FILE)
+        with mock.patch('__builtin__.open', mock_proto, create=True):
+            base_dirs = self._TASK._extract_base_dirs(
+                os.path.join('a', 'test', 'path', 'to', 'google', 'service',
+                             'v1', 'a.proto'))
+            repeated_base_dirs = self._TASK._extract_base_dirs(
+                os.path.join('home', 'foo', 'workspaces', 'google', 'service',
+                             'v1', 'googleapis', 'google', 'service', 'v1',
+                             'a.proto'))
```
Comment (jmuk) [inline test/test_python_package_change_task.py]: Oops, sorry I suggested a wrong test case. What about `/home/foo/workspaces/google/some/v1/googleapis/google/service/v1/a.proto`?  I want to make sure it skips `google/some/v1` part.

File: pipeline/tasks/gapic_tasks.py [hunk 8]
```diff
@@ -96,9 +95,8 @@ def execute(self, language, toolkit_path, descriptor_set, service_yaml,
             '--descriptor_set=' + os.path.abspath(descriptor_set), '--output='
             + os.path.abspath(code_root)
         ] + service_args + gapic_args
-        clargs = '-Pclargs=' + ','.join(args)
-        self.exec_command([os.path.join(toolkit_path, 'gradlew'), '-p',
-                           toolkit_path, 'runVGen', clargs])
+        self.exec_command(
+            task_utils.gradle_task(toolkit_path, 'runVGen', args))
```
Comment (garrettjonesgoogle) [inline pipeline/tasks/gapic_tasks.py]: Could you switch this to call `runCodeGen` while you're touching it?

Comment (geigerj) [inline pipeline/tasks/gapic_tasks.py]: Done

File: test/test_python_package_change_task.py (hunk 7 above)
Comment (geigerj) [inline test/test_python_package_change_task.py]: Done

File: pipeline/tasks/protoc_tasks.py [hunk 9]
```diff
@@ -49,7 +52,7 @@ def grpc_out_param(self, output_dir):
 
     @property
     def proto_compiler(self):
-        return 'protoc'
```
Comment (geigerj) [inline pipeline/tasks/protoc_tasks.py]: Done

**Pull request merged**